src/parser/SGTypes.ts

```diff
@@ -32,120 +164,128 @@ export class SGTag {
         this.setAttribute('id', value);
     }
 
+    /**
+     * Get the lower-case name of this tag.
+     */
+    public get tagName() {
+        return this.tokens.startTagName?.text;
+    }
+
+    /**
+     * Find all direct children by their tag name (case insensitive).
+     * This does not step into children's children.
+     *
+     */
+    public getChildNodesByTagName<T extends SGTag>(tagName: string) {
+        const result = [] as T[];
+        const lowerTagName = tagName.toLowerCase();
+        for (const el of this.childNodes) {
+            if (el.tokens.startTagName.text.toLowerCase() === lowerTagName) {
+                result.push(el as T);
+            }
+        }
+        return result as Readonly<Array<T>>;
+    }
+
+    /**
+     * Add a child to the end of the children array
+     */
+    public addChild<T extends SGTag>(tag: T) {
+        this.childNodes.push(tag);
+        return tag;
+    }
+
+    /**
+     * Remove a child from the children array
+     */
+    public removeChild(tag: SGTag) {
+        const idx = this.childNodes.indexOf(tag);
+        if (idx > -1) {
+            this.childNodes.splice(idx, 1);
+        }
+    }
+
     getAttribute(name: string): SGAttribute | undefined {
-        return this.attributes.find(att => att.key.text.toLowerCase() === name);
+        return this.attributes.find(att => att.tokens.key.text.toLowerCase() === name);
     }
 
     getAttributeValue(name: string): string | undefined {
-        return this.getAttribute(name)?.value?.text;
+        return this.getAttribute(name)?.tokens.value?.text;
```

Comment:
Why not using the getter here?

Reply:
Getters are technically function calls in javascript, so I figured this would be slightly more efficient.